### ISSUE TITLE
RFC: Cross-Cluster Data Migration

### ISSUE THREAD

**Issue by rohit-nayak-ps** [vitessio/vitess#7545]
## Motivation

Moving some or all tables from an external database or another Vitess cluster is currently not supported by the standard vreplication workflows. MoveTables can migrate data from one keyspace to another, but both keyspaces need to be in the same Vitess Cluster. We do have the ability to import tables from an external MySQL database, but it is difficult to set up.

There are use-cases where we want to move tables (or possibly the entire database) from one cluster to another. Either to adopt Vitess or to create a copy of an existing database. 

We propose creating a model similar to how unix allows copying files and folders from external filesystems, like:

```
Mount -t [mysql/vitess] <specific options> MyClusterName

Migrate -source MyClusterName.Database Create <TargetKeyspace.WorkflowName>
```

It will be possible to monitor/control the workflows using the current Workflow commands. Imports will internally create a MoveTables-like workflow and continue to replicate from the source until Complete is called.  \

```
Migrate Complete <TargetKeyspace.WorkflowName>
```

## Why Not MoveTables?

MoveTables has separate semantics than Migrate. Essentially Migrate is intended as a one-way copy whereas MoveTables allows you to serve either data from the source or target keyspace and switch between each other until you finalize MoveTables.

*   MoveTables sets up routing rules so that Vitess routes queries to the Source keyspace until a cutover
*   While switching Write traffic, in MoveTables, is possible to set up a reverse replication workflow so that the Source can be in sync with the Target, allowing you to revert back to the Source.  
However this requires that the Target can create vreplication streams (in the _vt database) on the Source database. This may not be always possible, for example, if the Source is a production system.
*   In MoveTables the tables already exist, just in a different keyspace. So the VSchema already contains these tables. While migrating, these tables will be available only after the Migrate is completed
*   Switching traffic is not meaningful in the case of Migrate since there is no query traffic to the original tables, as the Source is in a different cluster.

## Specifying Cluster Configuration

The Mount command will take a set of parameters which will depend on the type of cluster being defined. We will start with a basic set of config requirements for the POC and add security features afterwards. Credentials for a cluster will be specified while Mount-ing the cluster in the target cluster.  Certificate-based credentials will expect that the certificates will be local to the target cluster. 

### External Cluster Specs

The initial version will take topo parameters and pass them directly while instantiating a topo. The tablet info returned from the topo will be used to initiate gRPC connections to the source tablets. Once there is more clarity about cross-cluster security requirements we can enhance this.

**Topology**

*  <strong><code>topo_type</code></strong>  Topology types supported by Vitess: one of etcd2, zookeeper, consul
*  <strong><code>topo_server</code></strong> URL/port of the server
*  <strong><code>topo_root</code></strong> root node of the cluster in the topo

<strong>VTTablet GRPC</strong>

The POC will use the tablet connection params specified in the topo.

### External Database Specs

Parameters required to connect to the mysql server

*   <strong><code>socket </strong></code>
*   <strong><code>database</strong></code>
*   <strong><code>app_user</strong></code>
*   <strong><code>dba_user</strong></code>

## Implementation

MoveTables already contains all the required functionality of Migrate. We will be modifying this to create different code paths for Migrate. SwitchTraffic and ReverseTraffic will not be valid for a Migrate workflow. There will be no Reverse replication options. 

## Follow Up Issues

*   We might want to redact or munge certain pieces of data either for PII or for securing financially sensitive information. This is being planned in a follow-up feature for which an RFC will be available soon.

**Comment by jmoldow** [comment]
> In MoveTables the tables already exist, just in a different keyspace. So the VSchema already contains these tables. While migrating, these tables will be available only after the Migrate is completed

Why does this need to be different? Why can't you also require that the tables exist in the destination keyspace prior to starting the Migrate?

**Comment by jmoldow** [comment]
> Switching traffic is not meaningful in the case of Migrate since there is no query traffic to the original tables, as the Source is in a different cluster.

Maybe switching traffic isn't meaningful, but could you consider it meaningful to allow or deny traffic?

You say that the destination keyspace doesn't contain the tables, up until the point where the migration completes. Effectively, that means that traffic will be rejected, until completion.

Right before you create the tables on the destination vschema (allowing traffic to the destination), you could start denying traffic on the source. When traffic is denied, you can return a specific error code.

That way, an application system has a straightforward path of switching from one cluster to the other, without introducing corruptions at switchover time. It can issue dual queries to both clusters. At most one will allow the query, and the other(s) will reject the query. This allows the switchover to be controlled solely and atomically by the VReplication Workflow, rather than non-atomically by both the Workflow and the application.

All of this would apply only when you are actually switching application traffic from one to another, and when both the source is another Vitess cluster.


> While switching Write traffic, in MoveTables, is possible to set up a reverse replication workflow so that the Source can be in sync with the Target, allowing you to revert back to the Source.
However this requires that the Target can create vreplication streams (in the _vt database) on the Source database. This may not be always possible, for example, if the Source is a production system.

If the Source is a production system, and you are doing a Write switchover, doesn't that imply that the Destination must also be a production system?

Using my idea above (of rejecting traffic on the source, and then allowing traffic on the destination), couldn't you then use that window when traffic is rejected on both, to set up the reverse VReplication? Assuming the source and destination are both Vitess clusters.

I don't know if it's possible/safe for a vtctl command in one Vitess cluster to also talk to another Vitess cluster (which might be many versions ahead or behind). If not, maybe this can be performed by two commands, one to each cluster. The first command, to the source cluster, would start denying the traffic, and set up the reverse workflow. The second command, to the destination cluster, would start accepting the traffic, and delete the forward workflow.

**Comment by shlomi-noach** [comment]
> Right before you create the tables on the destination vschema (allowing traffic to the destination), you could start denying traffic on the source. When traffic is denied, you can return a specific error code.

In my understanding you may not have any form on control over the source. It may be a Vitess cluster, but it may also be just a MySQL server, or an Amazon Aurora setup etc.

**Comment by rohit-nayak-ps** [comment]
> > In MoveTables the tables already exist, just in a different keyspace. So the VSchema already contains these tables. While migrating, these tables will be available only after the Migrate is completed
> Why does this need to be different? Why can't you also require that the tables exist in the destination keyspace prior to starting the Migrate? Also, isn't VReplication driven by the destination? How can it route rows to the correct shard, if the VSchema isn't defined yet?

Note that, in the case of the external cluster being a Vitess one, we are talking about two separate clusters.  vtgate can only serve queries to the keyspace/tables in its own cluster. So the app pointing to the target cluster can never access the source tables. These tables will be only be truly available to an app targeting this cluster once Migrate has completed its copy phase and the eventual consistency of a vreplication workflow has been reached. 

The Migrate functionality is  intended as a _one way_ *copy* of the data from another cluster. The use-cases include taking a copy of production for trying out performance-related schema changes, for backup purposes, or for a copy of the data into staging or reporting environments.

**Comment by rohit-nayak-ps** [comment]
Closed via #7546
